test/test_expanded_weights.py

```diff
@@ -0,0 +1,409 @@
+            return
+        batch_size = input.shape[0]
+        with freeze_rng_state():
+            # get per sample grads with ExpandedWeights context manager, calling .backward() twice
+            test_module = TestModule(module)
+            actual_res = call_for_per_sample_grads(test_module, batch_size, input).sum()
+            actual_res.backward()
+            actual_grads = []
+            for param in module.parameters():
+                actual_grads.append(param.grad_sample)
+                del param.grad_sample
+
+            # get per sample grads with a for loop, running over the input twice
+            expected_grads = []
+            for i in range(batch_size):
+                res = module(input[i].unsqueeze(0)).sum()
+                expected_grads.append(torch.autograd.grad(res, module.parameters(), torch.ones_like(res)))
+            expected_grads = tuple(torch.stack(grad) for grad in zip(*expected_grads))
+        assert [torch.allclose(actual, 2 * expected) for (actual, expected) in zip(actual_grads, expected_grads)]
+
+    def test_per_sample_api_failing(self):
+        module = nn.Linear(10, 10)
+        input = torch.randn(64, 10)
+        with self.assertRaisesRegex(RuntimeError, r"Module passed must be nn.Module"):
+            call_for_per_sample_grads("fail", 64, input)
+        with self.assertRaisesRegex(RuntimeError, r"Batch size passed must be an integer"):
+            call_for_per_sample_grads(module, 6.4, input)
+        with self.assertRaisesRegex(RuntimeError, r"Batch size must be positive"):
+            call_for_per_sample_grads(module, -64, input)
+        with self.assertRaisesRegex(RuntimeError, r"incorrect for multiple calls"):
+            loss = call_for_per_sample_grads(module, 64, input).sum()
+            loss.backward()  # populate grad_sample fields
+            call_for_per_sample_grads(module, 64, input)
+
+class ContextManagerTests(TestBase):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    @property
+    def constructor_args(self):
+        return self._get_arg('constructor_args', False)
+
+    def test_context_manager(self, test_case):
+        module = self.constructor(*self.constructor_args)
+        input = self._get_input()
+        if len(input.shape) == 0 or input.shape[0] == 0:
+            return
+        if self.constructor == torch.nn.Linear and len(input.shape) == 1:
+            return
+        test_case._do_test(module, input)
+
+    def test_context_manager_multiple_inputs(self, test_case):
+        module = self.constructor(*self.constructor_args)
+        input = self._get_input()
+        if len(input.shape) == 0 or input.shape[0] == 0:
+            return
+        if self.constructor == torch.nn.Linear and len(input.shape) == 1:
+            return
+        test_case._do_test_multi_input(module, input)
+
+# TODO: Once all of these use ModuleInfo, replace with ModuleInfo tests
```

Comment:
Comment more on this is using the legacy nn tests... and that we'll move them eventully

Reply:
Fixed